crates/services/consensus_module/poa/src/service_test/trigger_tests.rs

```diff
@@ -299,3 +317,144 @@ async fn interval_trigger_doesnt_react_to_full_txpool() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn interval_trigger_produces_blocks_in_the_future_when_time_is_lagging() {
+    // Given
+
+    let block_time = Duration::from_secs(10);
+    let offset = Duration::from_secs(1);
+    let mut ctx = DefaultContext::new(Config {
+        trigger: Trigger::Interval { block_time },
+        signer: SignMode::Key(test_signing_key()),
+        metrics: false,
+        ..Default::default()
+    });
+    ctx.status_sender.send_replace(Some(TxId::zeroed()));
+    let start_time = ctx.test_ctx.time.watch().now();
+
+    // When
+
+    // We produce three blocks without advancing real time.
+    time::sleep(block_time * 3 + offset).await;
+    let first_block_time = ctx.block_import.try_recv().unwrap().entity.header().time();
+    let second_block_time = ctx.block_import.try_recv().unwrap().entity.header().time();
+    let third_block_time = ctx.block_import.try_recv().unwrap().entity.header().time();
+
+    // Then
+
+    // We should only have produced the three blocks.
+    assert!(matches!(
+        ctx.block_import.try_recv(),
+        Err(broadcast::error::TryRecvError::Empty)
+    ));
+
+    // Even though real time is frozen, the blocks should advance into the future.
```

Comment:
I feel like these tests might be confusing domains. 

Maybe we could have tests for the `TestTime` to show that it works independent of "real time", but I feel like the point of the `GetTime` port is that there is no such thing as "real time".

Reply:
i.e. This test should just be
```rs
async fn interval_trigger__when_advance_time_three_blocks_worth_then_produces_three_blocks() {}
```
then there can be a test for `TestTime` that is
```rs
async fn now__whe_use_tokio_sleep_then_current_time_advances()
```

Reply:
> Maybe we could have tests for the TestTime to show that it works independent of "real time", but I feel like the point of the GetTime port is that there is no such thing as "real time".

I guess I could use clearer terminology here. I refer to the tokio time as the "real time", because that's what it is meant to represent, whereas the timestamps are a read from the system clock (which can advance and rewind independently of the tokio time even when running live).

Reply:
I could refer to "tokio time" instead of "real time" in the coments.

Reply:
> `async fn now__whe_use_tokio_sleep_then_current_time_advances()`

Unfortunately, this is not necessarily the case. Or perhaps I'm misunderstanding you. What would you like to test in this scenario?

Reply:
For reference, we just had a chat about the concept of time and what we'd like to achieve and arrived at some conclusions:

1. Time in software systems are essentially (at least) two different but related concepts, we can call them the **wall clock** and the **stop watch**. The **wall clock** gives us a comparable time stamp that is understandable by other systems, but doesn't make many other guarantees. It may rewind or fast forward, making it unreliable for timing execution. For example if I want to make a request to an external resource with a 5s timeout at the same time as the wall clock fast forwards, the request may instantly time out. The **stop watch** instead provides us with pretty accurate relative times, allowing us to delay or pause execution of certain tasks for a certain amount of seconds or milliseconds. However, the stop watch does not provide us with a timestamp that we can send to another system.

2. Our current system uses `Tai64` for the wall clock and `tokio::time` as the stop watch. We use both concepts for different parts of the current business logic. The `GetTime` interface introduced in this PR captures the **wall clock** concept, thus making timestamps testable. However, I have not added any stop watch related functionality in the trait because 1. that would be pretty complex and result in a very substantial change and 2. the time functionality provided by tokio is already controllable in our tests.

3. Ideally we would like to reduce these two to a single concept of time, captured in a single trait which can both return a monotonically increasing somewhat accurate timestamp and provide decent control over execution time. Doing this is non-trivial, and will require making tradeoffs between execution time predictability and timestamp accuracy. This would also alter the current system behavior.

As exciting as it would be to dig into 3. we agree that it is beyond the scope of what is reasonable to do right now given our other priorities.

Reply:
@MitchTurner please let me know if there's anything you'd like to add or change in the above summary.